### PR TITLE
AWS SDK v2

### LINKED ISSUE #2302: AWS SDK 2.0 for Java
A developer preview of the AWS SDK for Java has been made available with a number of internal changes that could help improve performance of the GeoTrellis Amazon S3 backend: 

- Non-blocking I/O
- Pluggable HTTP client
- HTTP client sharing

See: https://aws.amazon.com/blogs/developer/aws-sdk-for-java-2-0-developer-preview/

### LINKED ISSUE #2466: Optimise S3 interaction
We figured out that `S3` connections are not parallelized well in the Java `AWS SDK`. 
It was measured that on slow connections it makes sense to parallelize all connections by spawning more threads and splitting the request data into smaller chunks. The slower internet connection is the bigger difference is easier to notice. According to this fact we can investigate how it is possible to speed up already fast (but probably not enough) `S3` connections (how to speed up `getObject(request).getObjectContent` calls).

1. How we can use `Futures` and how it is possible to determine an optimal way to split data into chunks to parallelize everything using [setRange](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/GetObjectRequest.html) queries. Does this approach makes sense at all?
2. There is an interesting [TransferManager](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/transfer/TransferManager.html) API, which works faster (or should work faster as my tests were limited) but it makes a good highly parallelized downloads into _files_. The disadvantage of this API that it works only with _files_ (downloads data from `S3` into _files_). We can consider building an in-memory version of it and verifying that it makes sense and it indeed effects on objects downloads. There is already an issue in their repo: https://github.com/aws/aws-sdk-java/issues/893

It is a bit a small R&D issue to clarify AWS S3 API and to double check that we use it efficient.

### PR DESCRIPTION
**Pull request by moradology**
## Overview

This PR removes v1 of the AWS SDK in favor of v2 and removes dependence upon the (Geotrellis provided) `S3Client` wrapper which was used to ease tests via `MockS3Client`. Work in the tests remains to implement a `MockS3Client` that simply extends the (Amazon provided as of v2 of the SDK) `S3Client` interface.

Tests aren't yet ready for the show as work remains to actually implement a `MockS3Client` before this can be merged but things are compiling and *should* work as before

### Checklist

- [x] `docs/CHANGELOG.rst` updated, if necessary
- [x] Unit tests added for bug-fix or new feature
- [x] Test on EMR cluster

### Notes

Some semantic differences between v1 and v2 need to be explored. Most pressingly, with v1 it was fine practice to spawn clients as necessary. This is no longer true. As of v2, it is recommended that clients be reused throughout the life of an application. This PR hasn't attempted to characterize or fix any problems arising from spawning clients. See [this](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/BestPractices.md#reuse-sdk-client-if-possible) for more

Closes #2466
Closes #2302

### CODE REVIEW

File: s3/src/main/scala/geotrellis/spark/io/s3/package.scala [hunk 1]
```diff
@@ -26,4 +31,27 @@ package object s3 {
       .mkString("/")
 
   implicit class withSaveToS3Methods[K](rdd: RDD[(K, Array[Byte])]) extends SaveToS3Methods(rdd)
+
+  def s3ObjectExists(bucket: String, key: String, s3Client: S3Client): Boolean = {
```
Comment (echeipesh) [inline s3/src/main/scala/geotrellis/spark/io/s3/package.scala]: These should be extension methods on the AWS S3 client.
Also, it's possible and even common that `HEAD` request will not be available on a bucket. I've seen this happen with requester pays buckets.

Comment (moradology) [reply s3/src/main/scala/geotrellis/spark/io/s3/package.scala]: Looks like the `ListBucket` perms are sufficient to infer a failed `HEAD` request's cause: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html

File: project/Settings.scala [hunk 2]
```diff
@@ -273,14 +273,14 @@ object Settings {
     /** https://github.com/lucidworks/spark-solr/issues/179 */
     dependencyOverrides ++= {
       val deps = Seq(
-        "com.fasterxml.jackson.core" % "jackson-core" % "2.6.5",
-        "com.fasterxml.jackson.core" % "jackson-databind" % "2.6.5",
-        "com.fasterxml.jackson.core" % "jackson-annotations" % "2.6.5"
+        "com.fasterxml.jackson.core" % "jackson-core" % "2.6.7",
```
Comment (pomadchin) [inline project/Settings.scala]: This update needs to be double checked on EMR cluster; see https://github.com/lucidworks/spark-solr/issues/179

**Comment by moradology** [comment]
Probably worth looking through while reviewing/reading any of this code: https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md

**Comment by pomadchin** [comment]
- [ ] software.amazon.awssdk 2.5.49 [CQ 19892](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19892)
- [ ] com.fasterxml.jackson.core:jackson-databind:2.6.7 [PB CQ 19887](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19887)
- [ ] com.fasterxml.jackson.core:jackson-annotations:2.6.7 [PB CQ 19888](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19888)
- [ ] com.fasterxml.jackson.core:jackson-core:2.6.7 [PB CQ 19889](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19889)
- [ ] org.reactivestreams:reactive-streams:1.0.2 [PB CQ 19890](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19890)
- [ ] com.typesafe.netty:netty-reactive-streams-http:2.0.0 [CQ 19893](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19893)
- [ ] io.netty:netty-all:4.1.33.Final [CQ 19895](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19895)

**Comment by pomadchin** [comment]
There are still some exceptions happen during the ingest:
```
19/05/24 17:14:29 ERROR Executor: Exception in task 802.0 in stage 9.0 (TID 20854)
software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Read timed out
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:97)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage$RetryExecutor.handleThrownException(RetryableStage.java:136)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage$RetryExecutor.execute(RetryableStage.java:94)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:62)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:57)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:37)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)
	at software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:240)
	at software.amazon.awssdk.core.client.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:96)
	at software.amazon.awssdk.core.client.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:120)
	at software.amazon.awssdk.core.client.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:73)
	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:44)
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:55)
	at software.amazon.awssdk.services.s3.DefaultS3Client.putObject(DefaultS3Client.java:3053)
	at geotrellis.spark.io.s3.S3RDDWriter$$anonfun$update$1$$anonfun$geotrellis$spark$io$s3$S3RDDWriter$$anonfun$$retire$1$1.apply(S3RDDWriter.scala:139)
	at geotrellis.spark.io.s3.S3RDDWriter$$anonfun$update$1$$anonfun$geotrellis$spark$io$s3$S3RDDWriter$$anonfun$$retire$1$1.apply(S3RDDWriter.scala:139)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:87)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
	at sun.security.ssl.InputRecord.read(InputRecord.java:503)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:975)
	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:933)
	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:282)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:138)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
	at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:165)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:185)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at software.amazon.awssdk.http.apache.internal.impl.ApacheSdkHttpClient.execute(ApacheSdkHttpClient.java:72)
	at software.amazon.awssdk.http.apache.ApacheHttpClient.execute(ApacheHttpClient.java:233)
	at software.amazon.awssdk.http.apache.ApacheHttpClient.access$500(ApacheHttpClient.java:102)
	at software.amazon.awssdk.http.apache.ApacheHttpClient$1.call(ApacheHttpClient.java:214)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.executeHttpRequest(MakeHttpRequestStage.java:66)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:51)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:35)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:64)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:77)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:39)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage$RetryExecutor.doExecute(RetryableStage.java:113)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage$RetryExecutor.execute(RetryableStage.java:86)
	... 29 more
```

**Comment by pomadchin** [comment]
It looks acceptable though; I would say it can be merged, since the job finished successfully, but I think requires some extra tuning (working on it right now): 

i3x-ingest-sdkv2

**Comment by pomadchin** [comment]
@moradology is 98f3656 tested on EMR?

**Pull request merged**